fix: allow for selecting patches instead of input files, seperate test logic from main

lw3266 opened the pull request:
known issue: when adding more input options to patch_to_input_mapping, coccinelle does not run the patch for some reason, for now I would like to know if there are any comments for how the test looks

currently you can run two options: aoc-coccinelle and aoc-coccinelle-test

both will allow you to select from a list of patches; aoc-coccinelle seems to work, and you can play with it by specifying a folder containing c source files and select all of the patches so it will write the combined result to a file, but I haven't throughly tested everything for aoc-coccinelle yet,

for now, my emphasis will be on the testing mechanism, and let me know if tehre is anything

lw3266 committed:
deletion

lw3266 committed:
Add files via upload

lw3266 commented:
@renatav 

> You are correct, the goal is to make sure that the current cocci files are correct. Adding a testing framework is a big steps towards that goal. When building a testing framework, the goal is to make sure that the code we are testing (which in this case are these cocci) files, is the same as the code that's being used in production. So, we want to be able to use the same cocci files within the testing framework, and to parse any C project. There should be no mapping between a cocci file and a test outside or the testing framework. You could use such a mapping to implement the tests, just make sure that you can accomplish that without copying the files. Would it help if I reorganized the code and added a very simple test?

Sorry I didn't see this when I was attempting to commit into my branch, and I ended up making a new branch in my forked repo and deleted the orginally files, hope that doesn't cause too much confusion to the time line.

Feel free to reorganize the code and try the tests, or let me know what to do, except that it might take me a little longer to do if that is you are ok with that,

Note that PR #21 intends to address the testing mechanism, so quite a few things has changed from #20

lw3266 committed:
Update README.md

lw3266 committed:
Update README.md

renatav reviewed (commented):
This is much better now. One thing that still needs to be addressed before this is merged in is the definition and comparison of expected and actual outputs. While we want to compare the results with the Clojure results, we do not want to embed this into the testing framework.   We want to check if actual and expected files are exactly the same (or if all matches are the same, ignoring the order). The expected files should be manually verified. Clojure's results can be used to update the expected output. It's possible that Clojure is not detecting something we want to detect, so its output should be manually validated as well.

lw3266 commented:
> This is much better now. One thing that still needs to be addressed before this is merged in is the definition and comparison of expected and actual outputs. While we want to compare the results with the Clojure results, we do not want to embed this into the testing framework. We want to check if actual and expected files are exactly the same (or if all matches are the same, ignoring the order). The expected files should be manually verified. Clojure's results can be used to update the expected output. It's possible that Clojure is not detecting something we want to detect, so its output should be manually validated as well.

I see, and the reason I hesitant to compare by the exact source is because I was unsure about the formatting of coccinelle output, due to indentations or formatting,

for instance, if one of the atom items is missing, and I attempt to add the missing item into the expected output file. Suppose we fix the patch to include that item, and the results are misaligned by one white space (there are indeed an leading white space for some of the captured items), 

while I understand your comment about Clojure, and here indeed I try to only use clojure output as a double check incase I missed something myself; and I actually copy and paste the coccinelle output for all of my expected output first, then manually check each one, so that the source actually exactly align for most of the cells,

I would like to hear your thoughts on this, and please correct me if I misunderstood your comment

lw3266 commented:
> > This is much better now. One thing that still needs to be addressed before this is merged in is the definition and comparison of expected and actual outputs. While we want to compare the results with the Clojure results, we do not want to embed this into the testing framework. We want to check if actual and expected files are exactly the same (or if all matches are the same, ignoring the order). The expected files should be manually verified. Clojure's results can be used to update the expected output. It's possible that Clojure is not detecting something we want to detect, so its output should be manually validated as well.
> 
> I see, and the reason I hesitant to compare by the exact source is because I was unsure about the formatting of coccinelle output, due to indentations or formatting,
> 
> for instance, if one of the atom items is missing, and I attempt to add the missing item into the expected output file. Suppose we fix the patch to include that item, and the results are misaligned by one white space (there are indeed an leading white space for some of the captured items),
> 
> while I understand your comment about Clojure, and here indeed I try to only use clojure output as a double check incase I missed something myself; and I actually copy and paste the coccinelle output for all of my expected output first, then manually check each one, so that the source actually exactly align for most of the cells,
> 
> I would like to hear your thoughts on this, and please correct me if I misunderstood your comment

lw3266 commented:
Actually, my previous comment regarding white spaces might be the best example as one can simple remove them,

However, take Omitted Curly braces for example, the brief practice I noticed from Coccinelle is that it will always capture the if statement, even if the else statement is missing the curly braces. 

In short, I found it difficult to predict the pattern that Coccinelle will use to capture source code, so I am unsure if I can create an exact match for the expected output

lw3266 committed:
Update test.py

comparing string content rather than line numbers

lw3266 committed:
Add files via upload

lw3266 committed:
Add files via upload

lw3266 committed:
Update README.md

lw3266 commented:
I tried my best to make the testing process work better. However, my current solution still requires a manual check in the end after all test cases are passed. Some of the thought process is in the README file.

Note that process.py is not used, and it will be removed in the future.

Currently, many of the tests will have some issue that mentions about float, rather than just a assertion error. This is due to the NaN values in the merged dataset that will create problem when turning into sets. I tried to dropna but encountered small issues unresolved. I am aware of this and will try to fix that soon.

lw3266 committed:
Update process.py

lw3266 committed:
Update test.py

lw3266 commented:
fix: small typo in the output debug file path; now NaN is properly filled with empty sets, there shouldn't be errors other than assertion errors, which are part of the testing mechanism. Added one line that allows for check word by word. ie: ['abc','cba'] == ['cba','abc'], ['ab','cba'] != ['cba','abc']

lw3266 committed:
Update test.py

lw3266 committed:
Update implicit_predicate.csv

fix: small error in expected data of unknown cause (possible due to autocorrect in my spreadsheet editor)

renatav commented on README.md:
```diff
@@ -0,0 +1,20 @@
+    -si "/stands for source-input/the path to folder of .c source files"
+    -v 0
+
+# The Expected Output
```
The top-level README file should not include information about how the testing framework works. There could be a `docs` folder with a `tests.md` file or something among those lines. If there are some remaining issues, it's better to create an issue

renatav commented on setup.py:
```diff
@@ -1,17 +1,22 @@
 setup( 
     name='atom-finder-coccinelle', 
-    version='0.0.1', 
+    version='0.1.2', 
```
This could be `0.1.0`

renatav commented on setup.py:
```diff
@@ -1,17 +1,22 @@
+    packages=['src', 'tests'],
     include_package_data=True,
-    install_requires=[],
+    install_requires=['pandas', 'click', 'pytest'],
```
I'm still leaning towards not using `pandas`. `pandas` is an overkill for small or simple data task. It's a rather big library with a number of additional dependencies, which could cause issues in the long run.

renatav commented on setup.py:
```diff
@@ -1,17 +1,22 @@
         'console_scripts': [
-            'atom-finder-coccinelle = main:cli',
+            "aoc-coccinelle = src.main:cli",
+            "aoc-coccinelle-test = tests.test:initialize_test"
```
When using `pytest`, you don't have to add another entry point

renatav commented on src/__init__.py:
```diff
@@ -0,0 +1 @@
+from src import main, option_select, run_cocci
```
It should be possible to avoid these imports

renatav commented on src/main.py:
```diff
@@ -0,0 +1,56 @@
+    required=True,
+    help="Directory containing your C source files.",
+)
+@click.option(
```
Your implementation of this looks fun from the developer's perspective, but from the usability standpoint, it's easier to be able to simply specify which patch you want to run directly when calling this command. On windows I use  `Cmder`, a console emulator, so I can use the `ctrl + R` shortcut to run previously executed commands. That makes it easy to e.g. run one patch against multiple files, or after making some fixes. You don't have to pick a patch every time you want to run just that one. By default, if not patch is specified, run all

renatav commented on src/main.py:
```diff
@@ -0,0 +1,56 @@
+    prompt='Choose an option to compare all or just a single file'
+)
+@click.option(
+    "-v",
```
This is not a high priority now, but the standard way of doing this would be to turn `-v` into a flag. So, if the command is called without `-v`, that means that more verbose logging is not necessary. having to specify 0 or 1 is not very user-friendly

renatav commented on src/main.py:
```diff
@@ -0,0 +1,56 @@
+    "--input",
+    type=Path,
+    required=True,
+    help="Directory containing your C source files.",
```
We should also be able to process individual files. If that Path points to a file, then we'll analyze that one file, if it's a directory path, then we want to run the patches against the whole directory.

renatav commented on src/main.py:
```diff
@@ -0,0 +1,56 @@
+        patch_list = [select(patch_list)]
+
+    input_list = []
+    pattern = ['*.c','*.cpp']
```
Make sure that this works recursively, if you have a directory inside a directory (that's pretty common)

renatav commented on src/main.py:
```diff
@@ -0,0 +1,56 @@
+    if not input.is_dir():
+        raise Exception(f"input path is not valid")
+
+    patch_list = (Path.cwd() / "cocci").glob("*.cocci")
```
The user could call the command from outside of the project, so you can't use `cwd()` here. Take a look at how using `__file__`  can be of help here

renatav commented on tests/__init__.py:
```diff
@@ -0,0 +1 @@
+from tests import test
```
This too is probably not necessary

renatav commented on tests/process.py:
```diff
@@ -0,0 +1,44 @@
+def compare_string_sets(strings1, strings2):
+    set1 = set(strings1)
+    set2 = set(strings2) 
+    return (len(set1)==len(set2))&all(any(s1 in s2 for s1 in set1) for s2 in set2)
```
This looks pretty complex, so I am not sure exactly what this does (a docstring might be helpful here), but using `&` is not Pythonic

renatav commented on tests/process.py:
```diff
@@ -0,0 +1,44 @@
+        print(f"Processing file: {file.name}")
+        remove_whitespaces_in_column(file)
+
+if __name__ == "__main__":
```
If you need this for testing purposes, you could write a tests that tests this :)

lw3266 replied:
sorry for the confusion, I should have not uploaded that file, process.py is not necessary at all, it is only to provide a reference of what I done to preprocess the files as well some functions I have tested

renatav commented on tests/test.py:
```diff
@@ -0,0 +1,192 @@
+    "type_conversion.c": "type_conversion.csv",
+}
+
+patch_path = Path.cwd() / "cocci"
```
Using `cwd` should be avoided wherever possible. Again, you could use `__file__`

renatav commented on tests/test.py:
```diff
@@ -0,0 +1,192 @@
+
+        expected = expected_path / input_to_expected_mapping[input]
+        
+        output = read_csv(data, header=None) # turns StringIO result into dataframe
```
I think that this can all be done without pandas, but it's not that important

renatav commented on tests/test.py:
```diff
@@ -0,0 +1,192 @@
+
+### Modify this later
+patch_to_input_mapping = {
+    "assignment_as_value.cocci": ["assignment_as_value.c"],
```
Instead of defining these mappings, I'd use `fixtures` to define paths to `cocci` files and input and expected files and would pass them into the test function. There would be a bunch of fixtures, but that's fine. It's cleaner

renatav commented on tests/test.py:
```diff
@@ -0,0 +1,192 @@
+def test_change_of_literal_encoding(input):
+    patch = "change_of_literal_encoding.cocci"
+    result = compare(patch, input, 0)
+    assert(result == True)
```
`assert` could be directly inside compare. `is_equal = combined['Containment'].all()` `assert is_equal`

renatav commented on tests/test.py:
```diff
@@ -0,0 +1,192 @@
+    # "type_conversion.cocci": "test_type_conversion",
+}
+
+# patch file list from keyboard selection, call corresponding test functions (same name as patch) through args, append input files in for each pytest parameter,  
```
Tests should not be interactive. If they are interactive, you won't be able to run them on CI. Even if you want to keep that feature, you can (and should) define a function which has a path as an input parameter and then test that function. If you absolutely have to test some code which relies on user's input (which is not the case here), you can mock those interactions.

lw3266 committed:
Update test.py

need to remove pandas next

lw3266 committed:
Update setup.py
